### PR TITLE
✨ feat(cli): et clean

### LINKED ISSUE #317: Review and clean-up existing issues

### LINKED ISSUE #976: ✨ feat(et): add `eest clean` to remove all generated folders and files
Originally suggested in ethereum/execution-spec-tests#461.

This would delete any cache files, build artifacts and potentially `uv`'s `.venv` and `tox`'s `.tox`:
https://github.com/ethereum/execution-spec-tests/blob/99f17f761bd3b5b4ba2ad7672eabe98259533787/src/entry_points/eest_utils.py#L94-L108

Other possible additions could be:
- `~/.cache/ethereum-spec-evm-resolver/` which stores the download execution-specs branches for each configured fork.
- `~/.cache/uv/` although I'd hope that this is not necessary.

### PR DESCRIPTION
**Pull request by raxhvl**
## 🗒️ Description
Introduces `uv run et clean`.

## 🔗 Related Issues
closes #976.  

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

File: src/cli/et/commands/clean.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    Remove all generated files and directories from the repository.
+    ```
+    rm -rf .tox .cache .pytest_cache .mypy_cache fixtures build venv
+    ```
+    Note: The virtual environment is not removed by default.
+    """
+    # List of items to remove
+    items_to_remove = [".tox", ".pytest_cache", ".mypy_cache", "fixtures", "build", "site"]
+    if all:
+        items_to_remove.append("venv")
```
Comment (raxhvl) [inline src/cli/et/commands/clean.py:24]: Should this be [`.venv`](https://github.com/astral-sh/uv/blob/590294a0acc59c609aae82813b9ad9d22e7dbbbf/docs/pip/environments.md?plain=1#L12C85-L12C92)?

Comment (danceratopz) [reply src/cli/et/commands/clean.py:24]: Yup, this changed to `.venv` with the introduction of `uv` #777.

Comment (danceratopz) [reply src/cli/et/commands/clean.py:24]: I would consider moving `.tox` to `all` as well, btw. I very rarely delete `.tox`.

**Comment by danceratopz** [comment]
One thing I do delete every once in a while is the `__pycache__` folders. 

```
find src/ -name '__pycache__' -exec rm -rf '{}' \;
```

We could use `glob` for this.

@raxhvl, @spencer-tb what do you think?

File: src/cli/et/commands/clean.py (hunk 1 above)
Comment (spencer-tb) [reply src/cli/et/commands/clean.py:24]: I'm for moving `.tox` to all! Let's start it all fresh and clean 🧼

**Comment by spencer-tb** [comment]
> One thing I do delete every once in a while is the `__pycache__` folders.
> 
> ```
> find src/ -name '__pycache__' -exec rm -rf '{}' \;
> ```
> 
> We could use `glob` for this.
> 
> @raxhvl, @spencer-tb what do you think?

Definitely agree! Starting fresh without having to reclone for me would the ideal scenario for `et clean`

I think we should delete `cached_downloads/` - the fixtures that are downloaded from the `consume` command when the input is a url. We could also delete `.pyspelling_en.dict`.

I'm happy with `--all` deleting `.venv` too!

LGTM other than that! Thanks for implementing this @raxhvl - I've wanted this feature for while

**Comment by raxhvl** [comment]
Would we also need a `--dry-run` command to inspect whats happening?

**Comment by danceratopz** [comment]
> Would we also need a `--dry-run` command to inspect whats happening?

:+1: This would be wonderful tbh.

**Comment by danceratopz** [comment]
> I think we should delete `cached_downloads/` - the fixtures that are downloaded from the `consume` command when the input is a url.

Up for this, but I'd add it to `--all` :)

**Comment by raxhvl** [comment]
How does this look for default vs `--all`?

```python
  items_to_remove = [".pytest_cache", ".mypy_cache", "fixtures", "build", "site", "cached_downloads", ".pyspelling_en.dict"]

    if all:
         # Delete __pycache__ directories using glob
         pycache_dirs = glob.glob("**/__pycache__", recursive=True)
         items_to_remove.extend(pycache_dirs)
         
         # Environment
         items_to_remove.extend([".tox", ".venv"])
```

File: docs/library/cli/et.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+## ⬛ `et make`
+
+Groups several commands that generates project files for you. It includes:
+
+### 1. `et make test`
+
+Creates a new specification test file for an EIP.
+
+Usage: `et make test [OPTIONS]`
```
Comment (danceratopz) [inline docs/library/cli/et.md:13]: Could you take a stab at using the pattern applied in the `evm_bytes` doc to grab the docstrings from `et` and sub-commands and include them in the doc?

The output is not perfect, but it's the most sustainable way to maintain our docs, imo.

https://github.com/ethereum/execution-spec-tests/blob/5cf1f24b39ddef259470f58413337b81bda081a6/docs/library/cli/evm_bytes.md?plain=1#L1-L6

Comment (danceratopz) [reply docs/library/cli/et.md:13]: Here's the result:
https://ethereum.github.io/execution-spec-tests/main/library/cli/evm_bytes/

As I said, the result's not perfect, so if you prefer to keep it simple, that's also fine. But mid-term I'd like to find a solution that works well for this.

**Comment by danceratopz** [comment]
@raxhvl @spencer-tb one more location that could be helpful to add to `all`: The directory containing the cached `ethereum/execution-specs` forks from the `ethereum-spec-evm-resolver`, cf this value:
https://github.com/petertdavies/ethereum-spec-evm-resolver/blob/c68756230a27709e10426d5fbe4fa1b142a0b0ee/src/ethereum_spec_evm_resolver/main.py#L43-L45

File: docs/library/cli/et.md (hunk 2 above)
Comment (raxhvl) [reply docs/library/cli/et.md]: Done!

**Comment by raxhvl** [comment]
>The directory containing the cached `ethereum/execution-specs` forks from the `ethereum-spec-evm-resolver`.

Is it this dir `Path(platformdirs.user_cache_dir("ethereum-spec-evm-resolver"))`. I guess we will have to add `platformdirs`

File: src/cli/et/commands/clean.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                    if os.path.isdir(item):
+                        shutil.rmtree(item, ignore_errors=False)
+                    else:
+                        os.remove(item)
+                except PermissionError:
+                    click.echo(f"❌ Permission denied to remove {item}.")
+                except Exception as e:
+                    click.echo(f"❌ Failed to remove {item}: {e}")
+
+                click.echo("🧹 Cleanup complete!")
```
Comment (danceratopz) [inline src/cli/et/commands/clean.py:70]: This is a little verbose as-is :slightly_smiling_face:. How about?
```suggestion
    click.echo("🧹 Cleanup complete!")
```

But we could add a verbose option and report the files deleted here?

Comment (raxhvl) [reply src/cli/et/commands/clean.py:70]: This will echo "Cleanup complete!" on even dry run. 

re: verbose - Adding a dry run style output for `-v` flag.

**Review by danceratopz — COMMENTED** [review]
Looks good to me! As discussed, feel free to refactor the click CLI docs to use #985.

**Comment by raxhvl** [comment]
> Looks good to me! As discussed, feel free to refactor the click CLI docs to use #985.

Done!

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -56,6 +56,7 @@ Test fixtures for use by clients are available for each release on the [Github r
 - 🐞 Fix `Bytecode` class `__eq__` method ([#939](https://github.com/ethereum/execution-spec-tests/pull/939)).
 - 🔀 Update `pydantic` from 2.8.2 to 2.9.2 ([#960](https://github.com/ethereum/execution-spec-tests/pull/960)).
 - ✨ Add the `et make test` command, an interactive CLI that helps users create a new test module and function ([#950](https://github.com/ethereum/execution-spec-tests/pull/950)).
+- ✨ Add the `et clean` command to help clean generated files and directories from the repository ([#980](https://github.com/ethereum/execution-spec-tests/pull/980)).
```
Comment (danceratopz) [inline docs/CHANGELOG.md:59]:
```suggestion
- ✨ Add the `et clean` command that helps delete generated files and directories from the repository ([#980](https://github.com/ethereum/execution-spec-tests/pull/980)).
```

**Review by danceratopz — COMMENTED** [review]
Thanks @raxhvl! Nearly there :slightly_smiling_face: two minor comments below.

**Comment by raxhvl** [comment]
Here is the updated help (truncated):

```
file for Ethereum execution specifications. The user is prompted to select
  the type of test, the fork to use, and to provide the EIP number and name.
  Based on the inputs, a test file is created in the appropriate directory
  with a rendered template.

  Prompts:

  * Choose the type of test to generate (State or Blockchain)

  * Select the fork where this functionality was introduced

  * Enter the EIP number

  * Enter the EIP name

  Example:

      uv run et make test

Options:
  -h, --help  Show this message and exit.

  Further help: https://ethereum.github.io/execution-spec-tests/main/writing_tests/
```
Docs:

image

**Review by danceratopz — APPROVED** [review]
Thanks a lot for this @raxhvl. This will very useful to regular and new contributors alike!

**Pull request merged**